File: tran.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

require 'net/http'
require 'uri'
require 'json'
require 'pp'

SRC_LANG = 'auto'
TO_LANG  = 'uk'
URL      = 'http://translate.google.com/translate_a/t'

def retrieve(params)
  uri = URI.parse(URL)
  uri.query = URI.encode_www_form(params)
  http = Net::HTTP.new(uri.host)
  request = Net::HTTP::Get.new(uri.request_uri)
  request['Accept-Charset'] = 'utf-8'
  request['User-Agent'] = 'Mozilla/5.0'
  response = http.request(request)
  return response
end

def get_result(data)
  output = ''

  parse_item = lambda do |item, depth|
    if item.is_a? Array
      item.each do |nested_item|
        if nested_item.is_a? Array
          parse_item.call(nested_item, depth + 1)
        elsif nested_item.is_a? String
          if nested_item.length != 0
            output += "\t" * depth if depth > 0
            output += nested_item + "\n"
          end
        end
      end
    elsif item.is_a? String
      if item.length != 0
        output += "\t" * depth if depth > 0
        output += item + "\n"
      end
    end
  end

  parse_item.call(data, -1)
  # output = output.encode('UTF-8',:invalid=>:replace, :replace=>"?")
  return output
end

def main
  params = { :client => 't', :sl => SRC_LANG, :tl => TO_LANG }
  params[:text] = ARGV[0]
  data = JSON.parse(retrieve(params).body.gsub(/,{2,}/, ',').gsub(/\[,/,'[').gsub(/,\]/,']'))
  puts "Translation: %s > %s\n\n" % [ SRC_LANG, TO_LANG ]
  puts get_result(data)
end

main()

File: tran.py
#!/usr/bin/env python
# -*- coding: utf-8 -*- 

import httplib, urllib, argparse, re
import json
# import pprint

SRC_LANG = 'auto'
TO_LANG  = 'uk'
HOST     = 'translate.google.com'

def argparser():
    parser = argparse.ArgumentParser(description='Instant Translator')
    parser.add_argument('phrase', metavar='p', type=str,
                        help='phrase to translate')
    args = parser.parse_args()
    return args.phrase

def retrieve(host, path):
    request = httplib.HTTPConnection(host)
    request.request("GET", path, "", {'User-Agent' : 'Mozilla/5.0'})
    response = request.getresponse()
    if response.status > 200:
        return u'[["{}"],"ERROR"]'.format(response.reason)
    return response.read()

def parse_item(item, depth = 0):
    if type(item) is list:
        result = u''
        for sub_item in item:
            result += parse_item(sub_item, depth + 1)
        return result
    elif type(item) is unicode:
        if len(item) == 0:
            return u''
        return u"{}{}\n".format(u"\t" * depth, item)
    else:
        return u''

def main():
    phrase = argparser()
    path = 'translate_a/t?client=t&sl=' + SRC_LANG + '&tl=' + TO_LANG + '&' + urllib.urlencode({'text' : phrase})
    response = re.sub(',{2,}', ',', retrieve(HOST, "/" + path))
    response = response.replace('[,', '[')
    response = response.replace(',]', ']')
    try:
        # translated = eval(response)
        translated = json.loads(response)
        # pp = pprint.PrettyPrinter(indent=2)
        # pp.pprint(translated)
        result = ''
        for item in translated:
            if type(item) is unicode:
                result = u"Translation: {} > {}\n\n{}".format(item, TO_LANG, result)
                # break
            result += parse_item(item, -1)
        print result.encode('utf8')
    except RuntimeError as ex:
        print u'Something went wrong ({}): {}'.format(response, ex)

if __name__ == '__main__':
    main()
